docs/key-issues/multiple_architectures.md

```diff
@@ -0,0 +1,129 @@
+# Platforms with multiple CPU architectures
+
+In addition to any ABI requirements, a binary is compiled for a CPU
+architecture. That CPU architecture defines the CPU instructions that can be
+issued by the binary.
+
+Historically, it could be assumed that an executable or library would be
+compiled for a single CPU archicture. On the rare occasion that an operating
+system was available for mulitple CPU architectures, it became the
+responsibility of the user to find (or compile) a binary that was compiled for
+their host CPU architecture.
```

Comment:
I think this article focuses too much on what (from my experience; away from the mobile ecosystem) is a niche use-case. Linux is very commonly compiled on all kinds of architectures, and the simple fact is that the architecture _needs_ to be taken into account (see "target triple" also in the glossary).

The `universal2` wheels are an exception due to Apple bridging the gap of their architecture change.

What's a common collection of architectures a binary needs to support on Android?

Reply:
Answer in reverse order: On Android, it's common to need armeabi-v7a, arm64-v8a, x86 and x86-64. The former two are 32 and 64 bit ARM, respectively; that's primarily for usage on device, but on macOS you'll need those for the simulator as well. The x86 and x86-64 images are entirely for simulator usage.

As for the other point: architecture is indeed important - each "slice" in a fat binary needs to be for the correct architecture. Compiler triples in particular are one of the biggest headaches in supporting multi-architecture builds. After all - what is the compiler triple for macOS supporting x86 *and* arm64? FWIW, Python's build system essentially ignores the architecture part of the compiler triple once it works out it's on macOS; when building third party libraries, you are essentially forced into the "lipo" solution I mention below - build multiple single-platform binaries, and then merge them into a single binary.

However, writing this use case off as "niche" is essentially saying "we don't consider mobile or macOS platforms to be in scope". 

On mobile platforms, multiple architecture *is* the use case. True, mobile usage isn't a big part of Python right now - but I'd argue that's a "chicken and egg" problem - nobody ships binaries for mobile platforms, so nobody uses Python on mobile, so nobody ships binaries for mobile platforms. And it's not easy for end-users to resolve this themselves - due largely to the fact that Python's binary build tooling doesn't deal well with multiple architectures and cross platform builds. However, in terms of the broader market, phones and tablets have market penetration that dwarfs that of desktop platforms. I would argue we ignore mobile platforms out our peril.

As for macOS: macOS spent 2005-2012 in a the universal1 era; The universal2 era started in 2019, so it seems reasonable to assume it will persist until ~2026. However, in between those two times, there was the x86/x86-64 paring, and it is possible to have a binary that supports both of *those* architectures simultaneously. It is also possible (but uncommon) to have a "really fat" binary that supports PPC, x86, x86-64, and M1 in a single binary. M2 is architecture compatible with M1, but it's not inconceivable that M3+ may require another period of fat binaries.

I don't think the comparison with Linux is entirely reasonable. Linux is essentially the "historical" case in this paragraph - the assumption is that the end user will download code and compile it for *their* computer; if they distribute code to another user, that user will compile it for *their* computer. *My* computer only ever has a single architecture, so I only ever need to compile for *my* architecture; and you can generally assume that the end user has (or can easily obtain) a compiler for their architecture. The idea of distributing binaries is essentially alien to the general Linux experience; you get binaries from your operating system's repository, and manylinux has established a set of specific conditions to allow sharing of Python module binaries, but outside those limited subsets, you hit limitations *very* quickly if you try to share a binary between two users. 

It's distribution that is the key factor here. The Linux use case essentially doesn't involve distribution. This isn't the case for Windows and macOS - Windows and macOS users *expect* to be able to download binaries, have them "just work", and share those binaries with their co-worker. The same is true of mobile platforms. It's possible to live your entire life without ever downloading a compiler for Windows or macOS; and you *can't* install a compiler on your mobile device.

Reply:
Thanks for the detailed reply!

> However, writing this use case off as "niche" is essentially saying "we don't consider mobile or macOS platforms to be in scope".

This is entirely due to my ignorance of that space (even though I _did_ see your 2019 PyCon keynote), and I'm not trying to write it off, as niche or otherwise (which is what I tried to caveat with "but maybe I don't know the mobile story well enough. Perhaps you can enlighten me further?").

That being said, fat binaries still feel like a hack to me. If the metadata and binary distribution were sufficiently good, you could just distribute the binary that a given installation requires, rather than duplicating (or worse) the footprint & bandwidth because the metadata isn't good enough.

I'll grant that one might view linux as the exception here, but it's kinda exceptions all the way down. Windows didn't really change architecture for a while (they tried their level best to hide x86 vs x86-64 from the user), and I don't think they're endorsing fat builds now that win-arm64 rolls around. OSX made a big bet to change architectures (that seems to pay off), but I don't see them eager to repeat that soon (perhaps I'm wrong). Mobile is it's own potpourri of history that you know way better than I do.

In any case, from the POV of describing the problems of the current ecosystem (the stated goal of pypackaging-native), it seems to me that fat binaries are a symptom (that can/should be explained) of an underlying problem (e.g. bad metadata, lack of architecture introspection of installers, etc.), but I don't (yet?) see why they necessarily need to be part of the solution space at all.[^1]

[^1]: it's of course possible that my idealistic distribute-only-what-you-need runs into real world issues and fat binaries might still come out as the least bad trade-off. But we're a ways off from the ecosystem-level discussion getting into that sort of nitty gritty.

Reply:
> The idea of distributing binaries is essentially alien to the general Linux experience; you get binaries from your operating system's repository, and manylinux has established a set of specific conditions to allow sharing of Python module binaries, but outside those limited subsets,

The idea of binaries from an operating system's repository being a "limited subset" strikes me as somehow highly strange.

And it definitely doesn't take into account the many, many, many projects out there that provide their own debian and Fedora repositories, which are architecture-bound and contain single-architecture packages that don't really care what actual version of either distro you are running. Or else are just compiled for the latest stable release -- that works too.

This is effectively identical to the manylinux specification, except manylinux is a sort of distro repository that works on many distros, but not nearly as well as a .deb or .rpm work.

Linux distro repositories are the original appstore, it just so happens that there's less fragmentation for Android and iOS -- while both have many versions, there's only the two platforms, not several dozen, so it's easy to package for either one.

It's also not a concept that Windows has missed out on noticing. Chocolatey is justifiably popular, and Microsoft based winget on that, sort of. Both allow easily downloading and installing existing architecture-specific *.msi installers etc. based on whichever architecture you are running.

...

There's only one key distinction to make between appstores for:
- Apple 
- everything else

Linux and Android both require vendors (whether the vendor is a distro or a project does not matter) to compile once for each architecture, and upload multiple binaries. Those multiple binaries, when ingested into an appstore, are abstracted away into the marketing name (package name or whatever), users ask to install that, and they get whichever one the appstore determines is best for them.

Apple requires:
- vendors to compile a single binary that is twice as big
- users to download a single binary that is twice as big
- users to store a single binary that is twice as big

And frankly, none of it is necessary at all.  A `*.pkg` should either specify the architecture it is for, or contain multiple architectures with each architecture tagged as relevant. For "universal2" packages, the appstore could perform the slicing for you, and only deliver the bits you need repackaged into a new .pkg, or simply work the way the appstore work like appstores on other platforms -- just deliver the one you need.

There is theoretical value in hosting the .pkg as a single file that users download and click on, although the only reason this value exists is because Apple doesn't like people to have control of their systems and offering custom appstore indexes makes Apple lose control. Even in this case, however, the installation process of that package should strip out what you do not need and avoid burning disk space frivolously. Granted that network bandwidth has already been frivolously wasted. :(

> However, writing this use case off as "niche" is essentially saying "we don't consider mobile or macOS platforms to be in scope".

What about option 3:

mobile or macOS platforms are in scope, but mobile platforms other than Apple aren't the niche under discussion, and Apple can be simultaneously in scope but also be creating its own unnecessary problems which is grumble-worthy. :)

Reply:
I guess the takeaway from this is, this article feels to me like:
- it de-emphasizes the state of the art in the *normative* case of "an operating system platform where multiple CPU architectures are supported" (which is nearly all platforms that are recent or even not-so-recent)
- the Android approach is not novel to Android -- it's the state of the art on Linux and the BSDs, and predates Android, but that's being ignored
- the PR is for "requirements exposed by mobile platforms" but it talks about "multiple architectures" which isn't the same at all, and then spends quite a bit of time talking about non-mobile platforms as "rare" or "on occasion" (?)

Reply:
> That being said, fat binaries still feel like a hack to me. If the metadata and binary distribution were sufficiently good, you could just distribute the binary that a given installation requires, rather than duplicating (or worse) the footprint & bandwidth because the metadata isn't good enough.

I won't argue they're a hack - but FWIW, they're a moderately elegant one. It is 100% a tradeoff of disk space for user experience. As a macOS user, I can "just download a binary", and it will "Just Work". I don't need to know whether I have an x86-64 or M1 machine, and I can't download the "wrong" binary for my machine.
 
From a developer's POV, it's also pretty elegant... as long as you're in the macOS ecosystem. In simple cases,  replacing `clang` with `xcrun --sdk macosx clang` will generate a fat binary, and you're done. Compilations only start to emerge when you hit against code that doesn't have a macOS story baked in. For example, if a project has a makefile that decomposes the compiler triple and tries to determine something significant from the architecture; or expects a `-D` flag to set the architecture, you're going to have a Bad Time. For these projects, you usually need to fall back on building multiple versions of the library and then mashing them together with `lipo`. It's not impossible to do, it's just fiddly, and most projects don't have a good place for "multiple pass then mash together" build logic to live, so you end up with meta-makefiles or similar.

> I'll grant that one might view linux as the exception here, but it's kinda exceptions all the way down. Windows didn't really change architecture for a while (they tried their level best to hide x86 vs x86-64 from the user), and I don't think they're endorsing fat builds now that win-arm64 rolls around. OSX made a big bet to change architectures (that seems to pay off), but I don't see them eager to repeat that soon (perhaps I'm wrong). Mobile is it's own potpourri of history that you know way better than I do.

I wouldn't take bets on another big platform shift any time soon; but I *would* take bets on a new iteration of ARM. ARM64 is really "ARMv8", and "ARMv9" has been announced. Apple hasn't announced anything about ARMv9 (AFAIK), but the fat approach allows them to support both in a single binary. On iOS, in the recent past it was necessary to support ARMv6, ARMv7, ARMv7s and ARM64 in a single binary if you wanted to support all Apple-supported devices.

> In any case, from the POV of describing the problems of the current ecosystem (the stated goal of pypackaging-native), it seems to me that fat binaries are a symptom (that can/should be explained) of an underlying problem (e.g. bad metadata, lack of architecture introspection of installers, etc.), but I don't (yet?) see why they necessarily need to be part of the solution space at all.[1](#user-content-fn-1-be6d2ab6e89d02da1a0841cf288f03f8)
> 
> ## Footnotes
> 1. it's of course possible that my idealistic distribute-only-what-you-need runs into real world issues and fat binaries might still come out as the least bad trade-off. But I think we're a ways off from the ecosystem-level discussion getting into that sort of nitty gritty. [↩](#user-content-fnref-1-be6d2ab6e89d02da1a0841cf288f03f8)

I don't agree that fat binaries are a symptom of a problem; I'd argue they're a solution to a problem that prioritises user experience over disk space (and possibly original developer experience).

As for why I think this should be in scope: IMHO, its a distribution problem - but a different distribution to the one that pip, conda, et al have historically been concerned with. 

Pip et al have historically considered "distribution" as a problem of a developer distributing a Python library to a Python package repository so that an end user can download that library into their virtualenv/conda env etc. In that context, I agree that multi-architecture problems don't exist. If I have an M1 macOS machine, I download the arm64 binary, and all is good with the world. 

The complication then comes when I, as a user who downloaded the M1 machine, try to share *my* project with my colleague, who has an x86-64 macOS machine. My single-platform binaries won't work on their machine; but a fat universal2 binary will. 

You could argue that I'm "doing it wrong" - that I should be documenting a requirements file, and my colleague should download their own set of requirements appropriate for their platform. However, that requires that my colleague understand the full ecosystem for the code. I can't just throw something onto a thumb drive and ask them to run it. For want of a term to hang this on, lets call this "app distribution".

In the mobile case, this is the *only* distribution mode that is possible. If you write an app that tries to download from PyPI/conda-forge, it will be rejected by the App Store (the Apple one, anyway; Play store might be a bit more lenient).

Reply:
> > The idea of distributing binaries is essentially alien to the general Linux experience; you get binaries from your operating system's repository, and manylinux has established a set of specific conditions to allow sharing of Python module binaries, but outside those limited subsets,
> 
> The idea of binaries from an operating system's repository being a "limited subset" strikes me as somehow highly strange.

This is literally what the manylinux definition is. A manylinux package may only link against a handful of binaries that are guaranteed to exist.

> And it definitely doesn't take into account the many, many, many projects out there that provide their own debian and Fedora repositories, which are architecture-bound and contain single-architecture packages that don't really care what actual version of either distro you are running. Or else are just compiled for the latest stable release -- that works too.

Sure. However, at that point, you're defining compatibility with "Fedora" or "Ubuntu" rather than a more nebulous "Linux". That approach works for users of those platforms, and I've definitely seen projects do that. But users of Mint, Arch, Manjaro, or any other distributions are out of luck (or, at least, will have highly variable luck); and I can tell you from my experience maintaining BeeWare that this long tail exists, and is deeply disappointed when your binary doesn't support their chosen niche Linux distro.
 
> There's only one key distinction to make between appstores for:
> 
> * Apple
> * everything else
> 
> Linux and Android both require vendors (whether the vendor is a distro or a project does not matter) to compile once for each architecture, and upload multiple binaries. 

Incorrect. As of August 2021, Google Play store requires the upload of a single, multi-platform AAB. That AAB can be constructed from individual platform-specific APKs, but the artefact you upload *must* be multi-platform.

> And frankly, none of it is necessary at all.

Says you. Apple has made a tradeoff between disk space and user experience. As a macOS user, I have no complaints about this tradeoff.

> There is theoretical value in hosting the .pkg as a single file that users download and click on, although the only reason this value exists is because Apple doesn't like people to have control of their systems and offering custom appstore indexes makes Apple lose control. 

I won't dignify this sort of trolling with a response.

Reply:
> You could argue that I'm "doing it wrong" - that I should be documenting a requirements file, and my colleague should download their own set of requirements appropriate for their platform. However, that requires that my colleague understand the full ecosystem for the code. I can't just throw something onto a thumb drive and ask them to run it. For want of a term to hang this on, lets call this "app distribution".

I've made the [argument](https://discuss.python.org/t/python-packaging-strategy-discussion-part-1/22420/49) on discourse that python needs an integration layer -- which would be the place to solve that, such that your colleague has nothing to do, and you (ideally) not much more.

I get the appeal of the story "I just want to give my colleague something to run", but I'm not convinced that we cannot come up with a better story than just mashing arches together because we don't know what platform we might end up running on.

> In the mobile case, this is the only distribution mode that is possible. If you write an app that tries to download from PyPI/conda-forge, it will be rejected by the App Store (the Apple one, anyway; Play store might be a bit more lenient).

Is this a question of artifact signatures (also required on M1) or because nothing can get installed on iOS that hasn't been signed off by Apple? For the former we'd have an answer (well, a hack) at least.

On a metalevel, I am not very sympathetic to improving Apple's ecosystem on a volunteer basis in the face of very deliberate hurdles they place in front of developers (and I say that as someone who has done just that for rolling out osx-arm support for many packages in conda-forge, often way before the original libraries supported that).

I understand the centrality of this stance in their business model, but I don't feel the wider community owes Apple anything in that regard. I'd much rather let them have an inferior product (for lack of a vibrant ecosystem), than mis-design our approach due to a platform that's outright hostile to what we're trying to achieve (i.e. distribute apps/binaries to users)[^1].

> I don't agree that fat binaries are a symptom of a problem; I'd argue they're a solution to a problem that prioritises user experience over disk space (and possibly original developer experience).
>
> As for why I think this should be in scope: IMHO, its a distribution problem - but a different distribution to the one that pip, conda, et al have historically been concerned with.

Changing tack a bit, how do other language ecosystems deal with distribution on phones / iOS, and can we learn from them? The scope of packaging problems is gigantic[^2] even before we add better coverage of nascent distribution channels.

Perhaps it's my biases coming through, but so far my impression is that we might want to have a mobile-specific key-issue page, without having mobile-heavy-but-not-generally-applicable pages about cross-compilation & fat binaries.

[^1]: "but what about the poor users that don't know any better?" -- over time, they always have the choice to vote with their dollars

[^2]: already just form the POV of the for (for) predominantly non-mobile set of users, which are at least as important to keep as it is to reach new ones.

Reply:
> > In the mobile case, this is the only distribution mode that is possible. If you write an app that tries to download from PyPI/conda-forge, it will be rejected by the App Store (the Apple one, anyway; Play store might be a bit more lenient).
> 
> Is this a question of artifact signatures (also required on M1) or because nothing can get installed on iOS that hasn't been signed off by Apple? For the former we'd have an answer (well, a hack) at least.

It's not a signing issue (although those also exist); it's a "platform security" issue. iOS App Store review policies require that all code that will be executed by the app are either (a) shipped by the app, or (b) entered by the user. You can't download code from an independent source that hasn't been through the App Store vetting process. 

This is essentially a compromise that allows app developers to use scripting languages in their own app, but doesn't allow developers to do an end-run around the App Store review process. Without this restriction, an app could essentially replace it's own running content, and Apple would not be able to verify that the app wasn't malicious. 

You can certainly argue whether this is a good or bad thing, or whether App Store review actually catches malicious behavior; but regardless, it's the current state of play.

> I understand the centrality of this stance in their business model, but I don't feel the wider community owes Apple anything in that regard.

Well, that's certainly a position one could maintain. However, I'd suggest that declaring 30% of the worldwide mobile ecosystem and 15% of the worldwide desktop ecosystem (numbers which jump to 60% and 30% respectively in the US and similar western nations) as "out of scope" because Apple doesn't conform to one specific model of software distribution will have consequences for the Python ecosystem as a whole.
 
> Changing tack a bit, how do other language ecosystems deal with distribution on phones / iOS, and can we learn from them? 

I will admit that I haven't done a *huge* amount of research, so I might be missing something; but my broad impression is that "they don't". The iOS and Android ecosystems both have very strong biases to their own languages (Swift and Kotlin, respectively). 

It's probably worth some additional research to make sure we're not missing some obvious prior art; but I wouldn't be at all surprised to discover that we're in uncharted territory.
 
> Perhaps it's my biases coming through, but so far my impression is that we might want to have a mobile-specific key-issue page, without having mobile-heavy-but-not-generally-applicable pages about cross-compilation & fat binaries.

A counterargument (framed in hyperbolic terms) - why are the concerns of GPU compatibility, which are used by a bare handful of packages in one very specific programming niche, more important than the concerns of an ecosystem that is, by some measures, the majority of worldwide computing devices?

I would not for a second claim that GPU compatibility considerations aren't important - they absolutely are. But I'd suggest that they're just part of the rich tapestry of issues facing a binary compilation and packaging ecosystem. IMHO, cross compilation and multi-architecture support are 2 other aspects of that tapestry. I would suggest treating mobile computing as a niche does the platform a disservice - especially when the issues that are being raised *aren't* mobile specific - they're also present on desktop macOS.

Reply:
> > I understand the centrality of this stance in their business model, but I don't feel the wider community owes Apple anything in that regard.
> 
> Well, that's certainly a position one could maintain. [...]

I argue it's Apple declaring itself out of scope for many reasonable approaches. Other platform manage without their lockdowns, while Apple wants to have more control (and provides their users with a cleaner experience for it). Redhat and several others are investing a lot into the broader open source ecosystem (and are highly interoperable for it); in contrast, Apple barely does (aside from something critical to them like clang & swift).

Not that Apple's stance is morally of philosophically wrong, but it comes with consequences, and the world is not static. If Apple's approach does not support the next big thing[^1] and (say) Android does, market forces will see to it that they change tack. The key point is that (in my personal opinion), volunteer-driven organisations do not have to bend over backwards for the vagaries of such a vendor.[^2]

[^1]: In a way that whichever maintainers are willing/able to do _for_ Apple
[^2]: As an IMO similar example, nVidia is starting to pivot towards a more open(-source) approach because it became too costly for various reasons to maintain their previous stance (that made it very hard for anyone trying to work with CUDA, though many people did nevertheless because the performance gains made it worth it for them).

> This is essentially a compromise that allows app developers to use scripting languages in their own app, but doesn't allow developers to do an end-run around the App Store review process.

I don't see (for example) how their review process is engraved in stone tablets, or how it couldn't include approved channels like an official python integration layer (which is what Palantir's Foundry and Snowflake's Snowpark do when they let Python into their closed systems through a curated entry point).

Reply:
Apple aside...

> I will admit that I haven't done a _huge_ amount of research, so I might be missing something; but my broad impression is that "they don't". The iOS and Android ecosystems both have very strong biases to their own languages (Swift and Kotlin, respectively).

This is kind of my point. Don't get me wrong - I think the work you and others are doing is absolutely necessary for the future, but solving everything at once is really hard or even impossible (and that's just for existing problems of the majority case _today_). So while the mobile case needs to be solved (and we shouldn't gratuitously throw stones in that way), I think that the addition of "blazing new trails"[^1] on top of a huge overhaul is... a bit gratuitous.

[^1]: for non-mobile-native languages

> A counterargument (framed in hyperbolic terms) - why are the concerns of GPU compatibility, which are used by a bare handful of packages in one very specific programming niche, more important than the concerns of an ecosystem that is, by some measures, the majority of worldwide computing devices?

So I agree with this, but how does that calculus look if you look at the percentages of "current Python users", or "current Python library maintainers". This is what I meant with [paraphrasing] "solving problems problems for the existing users is at least as important (probably more) than reaching new ones".

But I get that we have very divergent viewpoints on this, and there's nothing wrong with that. But you can perhaps take me as a stand-in for the average python maintainer, who (in all likelihood) has never packaged anything for mobile before, and so far, the documents you've written (haven't checked the new commits yet) read like a parallel world to me. Maybe we can agree on a common foundation and then go in more detail about challenges (mobile vs. ...) in a separate section resp. page.

Reply:
> The key point is that (in my personal opinion), volunteer-driven organisations do not have to bend over backwards for the vagaries of such a vendor.

Ironically, this is almost exactly my position when it comes to Linux binary support. I cannot express the size of the headaches that I've had trying to support interoperable Linux binaries with AppImage and Flatpak, or even getting manylinux binaries to work reliably.

Volunteers don't *have* to do anything. But at some point, you have to go where the users are. 

> I don't see (for example) how their review process is engraved in stone tablets, 

It's not engraved in stone - but it's not really something open for community debate or consultation. The clause that allows scripting languages at all was, for example, an addition that was made in response to industry feedback. 

I guess I'd say that an architectural choice that is based on "we'll get Apple to change their App Store review policy" is... not a viable starting position. 

> or how it couldn't include approved channels like an official python integration layer (which is what Palantir's Foundry and Snowflake's Snowpark do when they let Python into their closed systems through a curated entry point).

This sounds like an approach that would address Apple's concerns at a *technical* level; however, I'm not sure it's likely from a political/business standpoint - at least, not without someone doing a high-level strategic play with Apple at a corporate level.

> So while the mobile case needs to be solved (and we shouldn't gratuitously throw stones in that way), I think that the addition of "blazing new trails" on top of a huge overhaul is... a bit gratuitous.

Fair; although I'd argue that it *isn't* a new trail - it's a trail that has prior art for macOS support. Mobile exposes some very specific manifestations of these problems, but they're not new (or mobile specific) problems.

> So I agree with this, but how does that calculus look if you look at the percentages of "current Python users", or "current Python library maintainers". This is what I meant with [paraphrasing] "solving problems problems for the existing users is at least as important (probably more) than reaching new ones".

Can't argue with this. I'm definitely working in the "build it and they will come" space. 
 
> But I get that we have very divergent viewpoints on this, and there's nothing wrong with that. But you can perhaps take me as a stand-in for the average python maintainer, who (in all likelihood) has never packaged anything for mobile before, and so far, the documents you've written (haven't checked the new commits yet) read like a parallel world to me. Maybe we can agree on a common foundation and then go in more detail about challenges (mobile vs. ...) in a separate section resp. page.

I guess what I'm not seeing is why you consider it a parallel world, or why it's mobile specific. These problems exist *today* for macOS on desktop, and it's only by ignoring a major use case (i.e., restricting yourself to the "user downloads all their own dependencies") that these problems can be avoided.

Reply:
> This is literally what the manylinux definition is. A manylinux package may only link against a handful of binaries that are guaranteed to exist.

Not sure what the manylinux definition is doing here.

You're saying that: "binaries from your operating system's repository, and manylinux" are a "limited subset" outside of which things don't work, but they're actually just the model by which it does work generically.

> Sure. However, at that point, you're defining compatibility with "Fedora" or "Ubuntu" rather than a more nebulous "Linux". That approach works for users of those platforms, and I've definitely seen projects do that. But users of Mint, Arch, Manjaro, or any other distributions are out of luck (or, at least, will have highly variable luck);

Well, no. That .deb repository works on Linux Mint too -- Mint is .deb compatible, besides which Mint is a respin of Ubuntu. Add an Arch-compatible repo, and that makes only 3 packaging formats you need to target to cover pretty much everything. The holdout is source-based distros and musl libc-based ones.

And none of that is an issue with Windows for WinGet. One format works on all Windows OSes, after selecting the correct arch-specific download in the appstore manifest.

> Incorrect. As of August 2021, Google Play store requires the upload of a single, multi-platform AAB. That AAB can be constructed from individual platform-specific APKs, but the artefact you upload must be multi-platform.

The difference between requiring that one .deb needs to be uploaded per architecture, and requiring that one APK must be uploaded per architecture "but also stick it inside of a single bundle file during the upload process, we'll split it up on our end" is not impactful.

It's still compiling once per architecture, and uploading multiple binaries.

> Says you. Apple has made a tradeoff between disk space and user experience. As a macOS user, I have no complaints about this tradeoff.

:shrug: This is a very first-world perspective. And again, it's not a trade-off between disk space and user experience. It's a trade-off between "both disk space and internet speed/availability" and "user experience for the specific subset of macOS users that 'try to share my project with my colleague'".

It produces e-waste by default, disadvantages people who live in parts of the world where the internet is slow/metered, and doesn't have any meaningful benefit to the vast majority of macOS users, who do not share development projects but only download content from an appstore. Apple claims to be very interested in being carbon-neutral, but what does this actually mean when they gratuitously waste resources?

This may be the world we live in, it may be something people in release management have to know about, but it's really not a solution or an ideal world, and if we can find something better we should.

> Well, that's certainly a position one could maintain. However, I'd suggest that declaring 30% of the worldwide mobile ecosystem and 15% of the worldwide desktop ecosystem (numbers which jump to 60% and 30% respectively in the US and similar western nations) as "out of scope" because Apple doesn't conform to one specific model of software distribution will have consequences for the Python ecosystem as a whole.

If the desktop ecosystem can just use "inferior" wheels that are x64-specific, or arm-specific, is it really declaring the platform out of scope? What are the percentages for people on desktop that actually make use of fat binaries in a way that isn't fulfilled by regular binaries which pip cleverly selects? Still the same 15% of all desktop users?

> So while the mobile case needs to be solved (and we shouldn't gratuitously throw stones in that way), I think that the addition of "blazing new trails"[1](https://github.com/pypackaging-native/pypackaging-native/pull/27#user-content-fn-1-2970fa6738cc0985f972fcfb368e10dc) on top of a huge overhaul is... a bit gratuitous.

This. 100%